mdp/linear_flows_online.py

```diff
@@ -0,0 +1,607 @@
+from builtins import str
+import mdp
+from mdp import numx
+from .linear_flows import FlowException, FlowExceptionCR, _sys, _traceback
+from collections import deque as _deque
+
+
+class OnlineFlowException(mdp.MDPException):
+    """Base class for exceptions in Flow subclasses."""
+    pass
+
+
+class OnlineFlow(mdp.Flow):
+    """An 'OnlineFlow' is a sequence of online/executable nodes that are trained and executed
+    together to form a more complex algorithm.  Input data is sent to the
+    first node and is successively processed by the subsequent nodes along
+    the sequence.
+
+    Using an online flow as opposed to handling manually a set of nodes has a
+    clear advantage: The general online flow implementation automatizes the
+    training (including supervised training and multiple training phases),
+    execution, and inverse execution (if defined) of the whole sequence.
+
+    OnlineFlow sequence can contain trained or non-trainable Nodes and optionally
+    a terminal trainable Node.
```

Comment:
From reverse engineering mdp :) Here is my understanding between a few terms (correct me if I am wrong):
trainable - This defines whether a node is trainable or not. That is, is_trainable() returns True.
training - This is actually a state of a trainable node. If training completes is_training() returns False. 

A "trainable" node implies that is_trainable() = True. It does not indicate the current state of the node, that is, the training may have been finished or not. is_training() = True or False.
A "non-trainable" node implies that is_trainable() returns False and so does is_training().
A "trained" node implies that is_trainable() returns True, but is_training() returns False. 

So, an online flow could be a sequence of 
(a) OnlineNodes, or
(b) a mix of OnlineNodes and "trained" or "non-trainable" nodes, or
(c) a mix of OnlineNodes/trained/non-trainable nodes and a terminal "trainable" node. 

An OnlineFlow trains all the nodes simultaneously for the given input sample x. 
So, node2 is training while node1 is still training (node1.is_training() = True)

Whereas a standard Flow trains nodes sequentially, that is, the second node is only trained after the first node's training has finished  (node1.is_training() = False)

I have tried to make sure that online nodes only "learn" when their train method is called. execute method does not change the state of the algorithm only train method does.